Implement streaming obj serialization

yanchith opened the pull request:
This implements streaming OBJ serialization into a
`std::io::Writer`. Functionality is triggered by a main menu "Export OBJ"
button. Only unused geometries are exported. For each mesh or element of
mesh-array, a new "o" tag is created in the resulting OBJ file. The object name
is the same as the `return_value_name` that would be displayed in the pipeline
window in most cases. For some edge cases however, only the raw stringified var
ident is outputted. If the exported mesh is a mesh-array element and it's index
is not zero, it is part of the name written after the "o" tag.

The commit also does a bit of minor cleanup elsewhere:

- Ensures the session never generates the same variable identifier
  twice (related to being confident that if a var name is present in the
  session, it does not reference a previously removed variable declaration),
- Removes some per-frame uses of the format! macro from `ui.rs`,
- Inlines one simple call to tinyfiledialogs from `ui.rs` to `lib.rs`.

yanchith commented:
While dreaming, I realized a bug I introduced here. Gonna fix it now. Stay tuned.

yanchith commented:
@ondrowan @janper I'd appreciate a review of commit https://github.com/subdgtl/HURBAN-Selector/pull/159/commits/668c15d97e91db904b8db791fa9fd9107abc7bdb at least. I tried to keep the fix as minimal as possible, delegating further work for FIXMEs. Comments in the commit explain the problem and the "solution".

janper commented:
> @ondrowan @janper I'd appreciate a review of commit [668c15d](https://github.com/subdgtl/HURBAN-Selector/commit/668c15d97e91db904b8db791fa9fd9107abc7bdb) at least. I tried to keep the fix as minimal as possible, delegating further work for FIXMEs. Comments in the commit explain the problem and the "solution".

I understand the issue, I understand the fix, I don't understand the code though.
Nevertheless, are you really worried we'd run out of  9,223,372,036,854,775,807 idents in one session? And if so, can't you use u128, which gives you 340,282,366,920,938,463,463,374,607,431,768,211,455 idents?

janper commented on src/ui.rs:
```diff
@@ -924,16 +919,26 @@ impl<'a> UiFrame<'a> {
 
-                ui.separator();
+                status.export_obj = ui.button(
+                    imgui::im_str!("Export OBJ"),
```
there is a convention, that labels of buttons which open a new dialog end with a triple dot, so
`Export OBJ...`

yanchith replied:
Another task. Another PR.

janper replied:
I only want you to add three dots to this string. What is the issue?

yanchith replied:
Nevermind, I'll do it in here, it is a small thing. I'll also do https://trello.com/c/Z7Rp0yLS/325-reorder-menu-buttons

janper commented on src/lib.rs:
```diff
@@ -584,15 +590,86 @@ pub fn init_and_run(options: Options) -> ! {
+                if menu_status.export_obj {
+                    if let Some(path) = tinyfiledialogs::save_file_dialog_with_filter(
+                        "Save",
+                        "export.obj",
```
This is a good filename when the current project hasn't been saved yet. The export usually suggests the same filename as the current file, just a different extension, so of the project is called `my_model.hurban`, the the suggested export file name should be `my_model.obj`.

janper commented on src/lib.rs:
```diff
@@ -584,15 +590,86 @@ pub fn init_and_run(options: Options) -> ! {
 
+                if menu_status.export_obj {
+                    if let Some(path) = tinyfiledialogs::save_file_dialog_with_filter(
+                        "Save",
```
Please, change to `Export OBJ`

yanchith commented:
> Nevertheless, are you really worried we'd run out of 9,223,372,036,854,775,807 idents in one session? 

Not in one session. They don't reset. Ever. Meaning if you send a save file to someone and they load it, they don't start from 0. Regardless, I am not worried u64 won't be enough, even then. This is just an edge case that can be handled easily, so I think we should at some point. Hence the FIXME.

> And if so, can't you use u128

Same principle. Practically we won't ever run out, but it would be nice if we were handling this correctly.

yanchith commented:
> Could you please add a notification, just like in the case of the Screenshot save?

There is one. You probably don't see it because of the system dialog freezing our app though.

yanchith commented:
> Should it be possible to export an empty OBJ? I prefer disabling the button but also displaying a notification that the file hasn't been saved because it would be empty, would be OK.

Ok, I can disable it.

yanchith commented:
> I think you already mentioned it in the code comments, but there is quite a nasty behavior that the exporter exports currently computed geometry, not the current pipeline. So if you disable autoexecute then you may very easily export outdated version of the geometry. I suggest either informing about this in the notification area or even better trigger pipeline execution (if needed?) prior to the export.

It is not only about autoexecute. The same happens with autoexecute and long running operation. Actually, this is kind of the expected behavior, but I agree it is not intuitive :smile: Where do you want to notify the user of this? I don't think the notification area is the right place, because the system dialog vs notifications interaction I mentioned above.

I also don't think triggering an execution without at least a prompt is a good idea. Also, it won't solve the situation with a long running operation.

I'd perhaps just mention it in the button's description.

UPDATE: Discussed offline. The export button will be disabled (and have a tooltip explaining why), if the interpreter is either busy, or hasn't been sent the latest program to execute.

yanchith commented:
@janper I believe I addressed all of your points. Each point is addressed by a separate commit, should you want to review in detail. Let me know if you have more requests.

janper commented:
I've tested all scenarios I could think of and it seems to work great! Thank you.

There is still one issue though - the export button is clicable and functional even when disabled.

janper commented:
I see you even made the warning appear when relevant. Nice!

yanchith commented:
> There is still one issue though - the export button is clicable and functional even when disabled.

Ooops. One sec :smile:
